### PR TITLE
Add autoencoders (VAE) training and evaluation

### PR DESCRIPTION
**Pull request by SamitHuang**
# What does this PR do?

Adds VAE training and evaluation

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/mindspore-lab/mindone/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the documentation with your changes? E.g. record bug fixes or new features in `What's New`. Here are the
      [documentation guidelines](https://github.com/mindspore-lab/mindcv/wiki/%E6%96%87%E6%A1%A3%E7%BC%96%E5%86%99%E8%A7%84%E8%8C%83)
- [x] Did you build and run the code without any errors?
- [ ] Did you report the running environment (NPU type/MS version) and performance in the doc? (better record it for data loading, model inference, or training tasks)
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@xxx

### CODE REVIEW

**Review by CaitinZhao — APPROVED** [review]

File: mindone/models/autoencoders/autoencoder_kl.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# Copyright 2022 Huawei Technologies Co., Ltd
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/autoencoder_kl.py:1]: 要不这个copyright不加了

File: mindone/models/autoencoders/autoencoder_kl.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        self.embed_dim = embed_dim
+        if colorize_nlabels is not None:
+            assert type(colorize_nlabels) == int
+            self.register_buffer("colorize", ms.ops.standard_normal(3, colorize_nlabels, 1, 1))
+        if monitor is not None:
+            self.monitor = monitor
+        if ckpt_path is not None:
+            self.init_from_ckpt(ckpt_path, ignore_keys=ignore_keys)
+
+        self.split = ops.Split(axis=1, output_num=2)
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/autoencoder_kl.py:57]: 是不是都换成function的算子

File: mindone/models/autoencoders/modules.py [hunk 3]
```diff
@@ -0,0 +1,508 @@
+# Copyright 2022 Huawei Technologies Co., Ltd
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/modules.py:1]: 同上

File: mindone/models/autoencoders/modules.py [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+from mindspore import nn, ops
+
+_logger = logging.getLogger(__name__)
+
+
+def divisible_by(num, den):
+    return (num % den) == 0
+
+
+def is_odd(n):
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/modules.py:30]: 好像有共用的判断版本的func?

File: mindone/models/autoencoders/modules.py [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+        time_pad = dilation * (time_kernel_size - 1) + (1 - stride)
+        height_pad = height_kernel_size // 2
+        width_pad = width_kernel_size // 2
+        self.time_pad = time_pad
+        self.time_causal_padding = (width_pad, width_pad, height_pad, height_pad, time_pad, 0)
+
+        stride = (stride, 1, 1)
+        dilation = (dilation, 1, 1)
+        # diff from torch: bias, pad_mode
+        self.conv = ms.nn.Conv3d(
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/modules.py:60]: nn.Conv3d

File: mindone/models/autoencoders/modules.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+            chan_in, chan_out, kernel_size, stride=stride, dilation=dilation, has_bias=True, pad_mode="valid", **kwargs
+        )
+
+    def construct(self, x):
+        # FIXME: check dynamic shape issue in graph mode
+        pad_mode = self.pad_mode if self.time_pad < x.shape[2] else "constant"
+
+        # nn.Pad can be more efficient but it doesn't support 5-dim padding currently.
+        x = ms.ops.pad(x, self.time_causal_padding, mode=pad_mode)
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/modules.py:69]: ops.pad

File: mindone/models/autoencoders/modules.py [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+
+        return self.conv(x)
+
+
+def nonlinearity(x, upcast=False):
+    # swish
+    ori_dtype = x.dtype
+    if upcast:
+        return x * (ops.Sigmoid()(x.astype(ms.float32))).astype(ori_dtype)
```
Comment (CaitinZhao) [inline mindone/models/autoencoders/modules.py:78]: ops.sigmoid

File: examples/autoencoders/ae/models/modules.py [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+    ):
+        super().__init__()
+        # if use_linear_attn: attn_type = "linear"
+        self.ch = ch
+        self.temb_ch = 0
+        self.num_resolutions = len(ch_mult)
+        self.num_res_blocks = num_res_blocks
+        self.resolution = resolution
+        self.in_channels = in_channels
+        self.dtype = dtype
```
Comment (hadipash) [inline examples/autoencoders/ae/models/modules.py:466]: I think adding an explicit `dtype` to each cell unnecessarily increases complexity. I believe a better way is to control types through AMP.
